sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -151,19 +153,52 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   /**
    * Saves the content of the [[DataFrame]] as the specified table.
    *
+   * In the case the table already exists, behavior of this function depends on the
+   * save mode, specified by the `mode` function (default to throwing an exception).
+   *
    * @since 1.4.0
    */
   def saveAsTable(tableName: String): Unit = {
-    val cmd =
-      CreateTableUsingAsSelect(
-        tableName,
-        source,
-        temporary = false,
-        partitioningColumns.map(_.toArray).getOrElse(Array.empty[String]),
-        mode,
-        extraOptions.toMap,
-        df.logicalPlan)
-    df.sqlContext.executePlan(cmd).toRdd
+    if (df.sqlContext.catalog.tableExists(tableName :: Nil)) {
+      if (partitioningColumns.isDefined) {
+        sys.error("Inserting data into an existing partitioned table is not yet supported.")
```

Comment:
Inserting into (dynamically) partitioned table is actually supported now. However, a corner case we don't handle well for now is inserting into a Hive metastore table with static partitions. We don't load newly created partitions back into Hive metastore.